### PR TITLE
XR: Add body tracking support.

### PR DESCRIPTION
**Pull request by Malcolmnixon**
This pull request adds XRBodyTracker and XRBodyModifier3D support for driving avatar bodies.

image

https://github.com/godotengine/godot/assets/1863707/88d5d49b-8b8a-4e4e-8015-d7a03769b27f

https://github.com/godotengine/godot/assets/1863707/bb0dbea5-1635-4315-94eb-dbe2312b4b0e

### CODE REVIEW

**Comment by dsnopek** [comment]
This looks awesome!

> * Should we introduce an XRModifier3D common base for the 3 different modifier nodes?
> * Should we introduce an XRTracker common base for the 3 different tracker data objects?

My inclination would be NOT to make any base classes. Base classes are part of the public API, and I don't see them sharing a base being very useful for developers. And, for us, we could put any shared functionality in an internal utility class that is called as needed. I could see `XRHandModifier3D` and `XRBodyModifier3D` sharing the joint mapping and skeleton driving code, but they could both call into the same utility class to do that (composition vs inheritance).

> Tip bones can be useful for attachment points:
> * Should we add "tip" bone to head?  Godot Humanoid doesn't have a standard, but we could use "HeadTip"
> * Should we add "tip/toes" bones to feet? Godot Humanoid doesn't have a standard, but we could use "LeftTest/RightToes"

Do we have data for these tips? If so, it makes sense to me to drive bones for them if they are present, even if there is no such bone on the Godot Humanoid skeleton.

Or, if we don't have data, the developer could just add their own "tip bones" on the skeleton and that shouldn't disrupt anything, right? Or, do you just mean standardizing on bone names to use for these "tip bones"?

**Comment by Malcolmnixon** [comment]
> This looks awesome!
> 
> > * Should we introduce an XRModifier3D common base for the 3 different modifier nodes?
> > * Should we introduce an XRTracker common base for the 3 different tracker data objects?
> 
> My inclination would be NOT to make any base classes. Base classes are part of the public API, and I don't see them sharing a base being very useful for developers. And, for us, we could put any shared functionality in an internal utility class that is called as needed. I could see `XRHandModifier3D` and `XRBodyModifier3D` sharing the joint mapping and skeleton driving code, but they could both call into the same utility class to do that (composition vs inheritance).
> 
> > Tip bones can be useful for attachment points:
> > 
> > * Should we add "tip" bone to head?  Godot Humanoid doesn't have a standard, but we could use "HeadTip"
> > * Should we add "tip/toes" bones to feet? Godot Humanoid doesn't have a standard, but we could use "LeftTest/RightToes"
> 
> Do we have data for these tips? If so, it makes sense to me to drive bones for them if they are present, even if there is no such bone on the Godot Humanoid skeleton.
> 
> Or, if we don't have data, the developer could just add their own "tip bones" on the skeleton and that shouldn't disrupt anything, right? Or, do you just mean standardizing on bone names to use for these "tip bones"?

I've added HeadTip, LeftToes, and RightToes joints. These only make sense to drive if the user is driving positions. If the skeleton has them but we don't drive then they'll still move as childed and can be used as bone attachment points.

**Comment by Malcolmnixon** [comment]
I added a "show_when_tracked" boolean to let the user decide whether visibility should be driven by the tracking state. When disabled the developer has control over visibility.

**Comment by Malcolmnixon** [comment]
Rebased the PR and attached the videos testing against the https://github.com/GodotVR/godot_openxr_vendors/pull/97 pull request.

File: scene/3d/xr_body_modifier_3d.cpp [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+		// Get the skeleton bone (skip if not found).
+		const int bone = bones[i];
+		if (bone == -1) {
+			continue;
+		}
+
+		// Find the parent skeleton-bone.
+		const int parent_bone = skeleton->get_bone_parent(bone);
+		if (parent_bone == -1) {
+			// If no parent skeleton-bone exists then drive this relative to palm joint.
```
Comment (m4gr3d) [inline scene/3d/xr_body_modifier_3d.cpp:271]: The comment needs to be updated since this is being driven relative to the root joint.

**Review by m4gr3d — APPROVED** [review]

**Comment by TokageItLab** [comment]
So, I have a question.

Is it right that the tips bone is defined to get and position of the tracker and is only needed to calculate the orientation of the parent joint of the tips, and that the actual Skeleton does not need the tips? - In short, will this modifier work correctly without the tips joints in Skeleton?  (I can't debug since I don't have a tracker of any kind currently sorry.)

If it works correctly, then this PR is no problem and I can approved as Animation team. However, it would be a problem if tips requires the actual joints that Skeleton has. Currently, there is no way to add bones to skeleton on Editor. Also, most asset store models do not have tips.

If so, this should have the ability to add virtual joints along with `enum BoneDirection direction { +X, +Y, +Z, -X, -Y, -Z, FromParent} + float length` or `Vector3 tips_offset` options such as parent to child orientation or selecting a specific axis, and the tips specification should be optional.

**Comment by dsnopek** [comment]
> Keep in mind that modifications by set_bone_pose conflict with AnimationPlayer/Tree, so this kind of module will must have a base class like #87888 in the future.

Ah, good to know about that PR! I wasn't aware of it previously. I'll take a look and see if `XRHandModifier3D` needs to be modified to match the API of `SkeletonModifier3D`.

> Is it right that the tips bone is defined to get and position of the tracker and is only needed to calculate the orientation of the parent joint of the tips, and that the actual Skeleton does not need the tips? - In short, will this modifier work correctly without the tips joints in Skeleton?

Yes, if the tip bones aren't in the skeleton it simply won't drive them, and everything should still work correctly.

**Comment by TokageItLab** [comment]
In PR https://github.com/godotengine/godot/pull/87888, the XRModifier target is renamed to external skeleton, but I think that will be unavoidable because external skeleton is strange in this PR.

So, please just check if the migration for XRHand is correct. If it is working correctly, this PR should theoretically be able to migrate later.

**Review by TokageItLab — APPROVED** [review]
As explanation I received, the compatibility between tips definition and Godot Humanoid seems be no need to worry for now.

**Review by dsnopek — APPROVED** [review]
I haven't had a chance to test this yet, but the code all looks great to me!

(There's only Fredia's note about an out-dated comment, but I don't see that as a reason to necessarily hold this back.)

File: scene/3d/xr_body_modifier_3d.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	NodePath target;
+	BitField<BodyUpdate> body_update = BODY_UPDATE_UPPER_BODY | BODY_UPDATE_LOWER_BODY | BODY_UPDATE_HANDS;
+	BoneUpdate bone_update = BONE_UPDATE_FULL;
+	bool show_when_tracked = true;
+	JointData joints[XRBodyTracker::JOINT_MAX];
+
+	Skeleton3D *get_skeleton();
+	void _get_joint_data();
+	void _update_skeleton();
+	void _tracker_changed(StringName p_tracker_name, const Ref<XRBodyTracker> &p_tracker);
```
Comment (AThousandShips) [inline scene/3d/xr_body_modifier_3d.h:96]:
```suggestion
	void _tracker_changed(const StringName &p_tracker_name, const Ref<XRBodyTracker> &p_tracker);
```

**Review by AThousandShips — APPROVED** [review]

File: scene/3d/xr_body_modifier_3d.cpp (hunk 1 above)
Comment (Malcolmnixon) [reply scene/3d/xr_body_modifier_3d.cpp]: Good catch - comment updated!

File: scene/3d/xr_body_modifier_3d.h (hunk 2 above)
Comment (Malcolmnixon) [reply scene/3d/xr_body_modifier_3d.h]: Another good catch - changed to const-reference in both header and source.

**Comment by akien-mga** [comment]
Thanks!

**Comment by JekSun97** [comment]
Suggested an interesting idea thanks to this PR - https://github.com/godotengine/godot-proposals/discussions/9202

**Comment by HeadClot** [comment]
Hey @Malcolmnixon got a few questions for you about this. 

1. Can I drive the upper body with VR Controllers + Headset. While the lower body is driven by Animation and IK? For example a walking animation for the lower body.
2. What trackers are compatible with this?

**Pull request merged**